src/mainwindow.cpp

```diff
@@ -69,93 +101,111 @@ void MainWindow::handleReloadPorts() {
     loadPortsAndSet();
 }
 
-inline bool MainWindow::tryOpen() {
-    if (!m_monitor->m_serialPort.isOpen()) {
-        m_monitor->openingFlag = true;
-        emit open();
-        while (m_monitor->openingFlag);
-        if (!m_monitor->m_serialPort.isOpen()) return false;
-    }
-    return true;
-}
-
-void MainWindow::handleMonitorToggle(bool checked) {
+void MainWindow::handleMonitorToggled(bool checked) {
     if (checked) {
-        if (tryOpen()) {
-            emit startMonitor();
-        }
+        startMonitor();
     } else {
         stopMonitor();
     }
 }
 
-void MainWindow::handlePlotter(bool checked) {
+void MainWindow::handleReadyRead() {
+    // the first pass generally has corrupted data, so clear the serial port's buffer
+    if (m_readFirstPass) {
+        m_readFirstPass = false;
+        m_serialPort.clear();
+        return;
+    }
+    QByteArray buf = m_serialPort.readAll();
+    if (buf.length() > 0) {
+        emit sendToWorker(buf);
+    }
+}
+
+void MainWindow::handlePlotterToggled(bool checked) {
     if (checked) {
         m_plotterView = new PlotterView;
         m_plotterView->move(x() + 10 + width(), y());
         m_plotterView->show();
+        worker->plotEnabled = true;
+        connect(worker, &Worker::plotPoint, m_plotterView, &PlotterView::plotPoint);
         connect(m_plotterView, &PlotterView::finished, ui->plotterButton, &QToolButton::setChecked);
-        connect(m_monitor, &Monitor::plotPoint, m_plotterView, &PlotterView::plotPoint);
     } else {
+        worker->plotEnabled = false;
         m_plotterView->close();
         delete m_plotterView;
         m_plotterView = nullptr;
     }
-    m_monitor->plotFlag = checked;
 }
 
 void MainWindow::handleSend() {
     if (ui->lineEdit->text().length() != 0 &&
             m_availablePorts.length() != 0 &&
             tryOpen()) {
-        m_monitor->m_serialPort.write(ui->lineEdit->text().toUtf8());
-        if (m_monitor->m_serialPort.error() == QSerialPort::WriteError || !m_monitor->m_serialPort.waitForBytesWritten(5000)) {
-            handleError("Failed to write to port");
+        m_serialPort.write(ui->lineEdit->text().toUtf8());
+        if (!m_serialPort.error()) {
+            ui->lineEdit->clear();
         }
-
-        ui->lineEdit->clear();
     }
 }
 
-void MainWindow::handleError(const QString& err, bool monitorReset, bool warning) {
-    if (monitorReset) {
+void MainWindow::handleError(QSerialPort::SerialPortError err) {
+    if (err != QSerialPort::NoError) {
         resetMonitor();
     }
-    if (warning) {
-        QMessageBox::warning(this, "Warning", err, QMessageBox::Ok);
-    } else {
-        QMessageBox::critical(this, "Error", err, QMessageBox::Ok);
+    switch(err) {
```

Comment:
Logging and error messaging should be moved somewhere else like in a `Log` class that can be configured at compile time with either `stderr`/`stdout`, `QMessageBox` or a UI component with idk maybe something like `-DLOGGING_MODE=qmessage`

Reply:
Well, so do we not want to inform the end user about an error unless they manually enable logging? Like, these errors can happen if suppose the user accidentally disconnects the board while it's in use, or if the board is not properly connected, so they are pretty common

Reply:
Well obviously there'd be a default value but there should be an option to pipe logging elsewhere